[PyTorch][Edge] Improve InflatableArgs for Bundled Inputs

pavithranrao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62368

# Context

The bundled inputs accepts an expression in the form of string `InflatableArg.fmt` that can be applied on the inputs to inflate. The `InflatableArg.fmt` provides flexibility to have custom transformation to inflate. When the input arguments to a function are not Tensor type, TorchScript casts the inputs from type `T` to `Optional[T]` expects the function to handle Nullable (None) clause as well. This becomes tricky to handle in one line code or lambda functions. 

We propose an alternative way which allows `InflatableArg` to include the text of a TorchScript function that would be defined on the module as a helper, then use that in its inflation expression. This can be provided by `InflatableArg.fmt_fn`. Please refer to `pytorch/test/test_bundled_inputs.py` for example on how to use the same.

Also refer @JacobSzwejbka's comment on the same https://github.com/pytorch/pytorch/pull/62368#issuecomment-892012812

# Mitigation
Allow InflatedArg to include the text of a TorchScript function that would be defined on the module as a helper, then use that in its inflation expression.

Differential Revision: [D29950421](https://our.internmc.facebook.com/intern/diff/D29950421/)

pavithranrao commented on test/test_bundled_inputs.py:
```diff
@@ -324,5 +324,96 @@ def forward(self, arg):
+        def condensed(t):
+            ret = torch.empty_like(t).flatten()[0].clone().expand(t.shape)
+            assert ret.storage().size() == 1
+            ret.storage()[0] = 0
```
mypy lint fails with the error:

```
+ mypy --config=mypy.ini
test/test_bundled_inputs.py:354:13: error: Unsupported target for indexed assignment ("Storage")  [index]
Found 1 error in 1 file (checked 1397 source files)
```

JacobSzwejbka commented on torch/utils/bundled_inputs.py:
```diff
@@ -240,6 +246,9 @@ def augment_many_model_functions_with_bundled_inputs(
     get_bundled_inputs_functions_and_info_template = ""
+    inflate_helper_names: List[str] = []
+
+    model._c._register_attribute("inflate_helper_names", ListType(StringType.get()), [])
```
nit: inflate_helper_names -> _ inflate_helper_names

JacobSzwejbka commented:
For context on the reason since the link is internal only. Some users wish to bundle complex types like Optional[Dict[foo, bar]]. The previous syntax for writing custom inflatable args already made this cumbersome to do, but types like Optional probably arent even supported at all. Expanding this to support functions makes the syntax of writing them much simpler (its just a torch script function like you would write for you model), but also makes what you can do way more powerful.

pavithranrao committed:
[PyTorch][Edge] Improve InflatableArgs for Bundled Inputs (#62368)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/62368

# Context
To know about the reason please read: https://fb.workplace.com/groups/1405155842844877/permalink/900106737234474/

# Mitigation
Allow InflatedArg to include the text of a TorchScript function that would be defined on the module as a helper, then use that in its inflation expression.
ghstack-source-id: 135158680

Test Plan:
To run `test_dict_args`

```
(base) [pavithran@devvm1803.vll0 /data/users/pavithran/fbsource/fbcode] buck test //caffe2/test:test_bundled_inputs -- test_dict_args
Action graph will be rebuilt because files have been added or removed.
Building: finished in 5.4 sec (100%) 12180/12180 jobs, 0/12180 updated
  Total time: 5.8 sec
More details at https://www.internalfb.com/intern/buck/build/fafcf277-1095-4cba-978d-6022f0d391ad
Tpx test run coordinator for Facebook. See https://fburl.com/tpx for details.
Running with tpx session id: 5ef9de71-c1b1-406b-a6c0-3321c2368b8d
Trace available for this run at /tmp/tpx-20210727-163946.454212/trace.log
Started reporting to test run: https://www.internalfb.com/intern/testinfra/testrun/7036874465805934
    ✓ ListingSuccess: caffe2/test:test_bundled_inputs - main (11.365)
    ✓ Pass: caffe2/test:test_bundled_inputs - test_dict_args (test_bundled_inputs.TestBundledInputs) (12.307)
Summary
  Pass: 1
  ListingSuccess: 1
If you need help understanding your runs, please follow the wiki: https://fburl.com/posting_in_tpx_users
Finished test run: https://www.internalfb.com/intern/testinfra/testrun/7036874465805934
```

To check the py code of TS module:
P433043973

Differential Revision: D29950421

fbshipit-source-id: 9466ef7b77efdae37d4381c36f894cca31f765f6
